rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/rename/Common.rsc

```diff
@@ -171,39 +194,116 @@ tuple[bool, bool](Tree) allNameSortsFilter(str name1, str name2) {
+            case (Nonterminal) `<Nonterminal n>`: for (nt1 <- escNonterminals, n := nt1) return true;
+            case (NonterminalLabel) `<NonterminalLabel n>`: for (ntl1 <- escNonterminalLabels, n := ntl1) return true;
+            case (QualifiedName) `<QualifiedName n>`: {
+                if (n.names[0].src == n.src) { // skip unqualified names
+                    for (qn <- qualifiedNames, qn := n || qn := [QualifiedName] normalizeEscaping("<n>")) return true;
+                }
+            }
+        }
+
+        return false;
+    };
+}
+
+bool(loc) containsQualifiedNameCheck(str _) =
+    bool(loc _) { return true; };
+
+bool(loc) containsNameCheck(str name) =
+    bool(loc l) {
+        return contains(readFile(l), name);
+    };
+
+@synopsis{
+    Filters a given set of locations based on occurrence of `name` in parse tree.
+    Accepts a loose pre-check builder function that, given a name, returns a function that, given a location, cheaply computes whether the parse tree for that location *might* contain an occurrence of the given name.
+}
+set[loc] filterFiles(set[loc] fs, str name, bool(loc l)(str) loosePreCheckBuilder, Tree(loc) getTree) {
+    j = "Checking files for occurrences of \'<name>\'";
     jobStart(j, totalWork = size(fs));
     set[loc] filteredFs = {};
+    treeFilter = singleNameFilter(name);
+    loosePreCheck = loosePreCheckBuilder(name);
     for (loc f <- fs) {
         jobStep(j, "<f>");
-        if (treeFilter(getTree(f))) filteredFs += f;
+        if (loosePreCheck(f) && treeFilter(getTree(f))) filteredFs += f;
     };
     jobEnd(j);
 
     return filteredFs;
 }
 
-tuple[set[loc], set[loc]] filterFiles(set[loc] fs, tuple[bool, bool](Tree) treeFilter, Tree(loc) getTree) {
+@synopsis{
+    Filters a given set of locations based on occurrences of `name1` and `name2` in parse tree.
+    Accepts a loose pre-check builder function that, given a name, returns a function that, given a location, cheaply computes whether the parse tree for that location *might* contain an occurrence of the given name.
+}
+tuple[set[loc], set[loc]] filterFiles(set[loc] fs, str name1, str name2, bool(loc l)(str) loosePreCheckBuilder, Tree(loc) getTree) {
     set[loc] fs1 = {};
     set[loc] fs2 = {};
 
-    j = "Checking files for occurrences of relevant names";
+    j = "Checking files for occurrences of \'<name1>\' and \'<name2>\'";
     jobStart(j, totalWork = size(fs));
+
+    twoNameTreeFilter = twoNameFilter(name1, name2);
+
+    loosePreCheck1 = loosePreCheckBuilder(name1);
+    loosePreCheck2 = loosePreCheckBuilder(name2);
+
+    singleNameTreeFilter1 = singleNameFilter(name1);
+    singleNameTreeFilter2 = singleNameFilter(name2);
+
     for (loc f <- fs) {
         jobStep(j, "<f>");
-        tr = getTree(f);
-        <l, r> = treeFilter(tr);
-        if (l) fs1 += f;
-        if (r) fs2 += f;
+        if (loosePreCheck1(f)) {
+            if (loosePreCheck2(f)) {
```

Comment:
This potentially reads the file twice, what if we do a read, get the string, and use that for both checks?

(There is even a parse function overload where you can pass in the string and the loc).

Reply:
Changed this to read the file once for all pre-checks. The parse overload is a bit tricky to use, since it does not make use of the caching by the framework. I think the file read times are dwarfed by actual parse times, so I am not sure this is an optimization we need to do now.